databuilder/query_model.py

```diff
@@ -0,0 +1,388 @@
+# is a mapping from colum names to types which allows us to check that only columns
+# which exist are being accessed and that all operations use valid types. The default
+# schema assumes that all referenced columns exist and gives them the type Any. This
+# obviously limits the validation we can do but it does not eliminate it. For instance,
+# we can tell when an operation which can only return a bool is used as an argument to
+# one which only accepts an int.
+class DefaultSchema:
+    def __init__(self, name):
+        self.name = name
+
+    def __getitem__(self, column_name):  # pragma: no cover
+        return Any
+
+    def __repr__(self):  # pragma: no cover
+        # Gives us `self == eval(repr(self))` as for dataclasses
+        return self.name
+
+
+DEFAULT_SCHEMA = DefaultSchema("DEFAULT_SCHEMA")
+
+
+# BASIC QUERY MODEL TYPES
+#
+# The Query Model consists of operations on "frames" and "series". A frame is a table-like
+# thing that has rows and columns, and a series is a column-like thing that contains a
+# sequence of values. Frames can be created from SQL tables, and then various filtering,
+# transformation and sorting operations can be applied to produce new frames. Likewise,
+# a series can be created from a SQL column and then transformed and combined with
+# others to produce new series.
+#
+# Central to the Query Model design is that these frames and series have an additional
+# property called their "dimension": do they contain at most one row per patient, or
+# might they contain many rows?
+
+
+class Node:
+    "Abstract base class for all objects in the Query Model"
+
+    def __init_subclass__(cls, **kwargs):
+        # All nodes in the query model are frozen dataclasses
+        dataclasses.dataclass(cls, frozen=True)
+
+    def __post_init__(self):
+        # validate the things which have to be checked dynamically
+        validate_node(self)
+
+
+class Frame(Node):
+    ...
+
+
+class Series(Node):
+    def __class_getitem__(cls, type_):
+        # Series have an "inner" type denoting the type of value they contain e.g
+        # `Series[int]`. This is the method which makes that syntax work. At the moment
+        # we don't do anything with this type, we just return the original class
+        # unmodified. Later we will handle this properly and validate the types.
+        return cls
+
+
+class OneRowPerPatientFrame(Frame):
+    ...
+
+
+class ManyRowsPerPatientFrame(Frame):
+    ...
+
+
+class OneRowPerPatientSeries(Series):
+    ...
+
+
+class ManyRowsPerPatientSeries(Series):
+    ...
+
+
+class SortedFrame(ManyRowsPerPatientFrame):
+    ...
+
+
+# OPERATIONS
+#
+# Operations indicate the kind of thing they return by subclassing one of the basic
+# types above.
+
+
+# `Value` wraps a single static value in a one-row-per-patient series which just has
+# that value for every patient. This simplifies the signature of all other operations
+# which no longer need to care about static values.
```

Comment:
Would it be correct to think that using `Value` is like [broadcasting][1] in numpy? It takes a scalar and stretches it over the number of rows in the one-row-per-patient series.

[1]: https://numpy.org/doc/stable/user/basics.broadcasting.html

Reply:
Sort of. Although that's also true of pretty much all the constructs here. And this stretching applies to multi-row-per-patient series as well as the one-row-per-patient ones, depending on the context.